Add Delft3D4-WEADS coupling framework

shabnamm7 opened the pull request:
This PR adds a grid-based coupling framework between Delft3D-FLOW and WEADS ecological model, following the WEADS interface. It includes:

- `WEADS_delft3d4.py`: Main driver script for coupling
- Annual coupling loop in `run_delft3d4_loop.sh`
- Preprocessing, MEM, and postprocessing modules in `src_point/*_delft3d4.py`
- Utility scripts in `utils/`, including water level extraction and plotting
- Sample input files in `test_delft3d4/`
- Conda environment file: `env_delft3d4.yml`
- Documentation in `README_delft3d4.md`

All additions follow a modular naming convention (suffix `_delft3d4`) to avoid conflicts with existing WEADS components.

jinikeda commented on src_point/postprocessing_delft3d4.py:
```diff
@@ -0,0 +1,136 @@
+        for row in updated_matrix:
+            for j, val in enumerate(row):
+                val_str = f'{val:.7E}'
+                spacing = '  ' if val < 0 else '   '
```
@shabnamm7, please carefully respond copilot's review first.

shabnamm7 commented:
I’ve removed the local delft3d_matlab folder and added it as a Git submodule pointing to https://github.com/shabnamm7/delft3d_matlab_scripts.
Also updated waterlevel_delft3d4.m to use a relative addpath for submodule compatibility.
Ready for review — thank you!

jinikeda commented:
@shabnamm7, Please first resolve the existing issues that I already requested to modify, including the model bug of "No such file or directory: 'water_level_extracted.csv'"

Also, source codes should be in the src_xxx folder, instead of 'utils'. Let's temporarily use the folder name "src_delft3d"
Don't use "test_delft3d4" for the input folder name. The tests folder should contain pytest codes as you worked on the CE7701 project (https://github.com/cekees/ce7701_2025).

Thanks for your help.

jinikeda commented:
@shabnamm7, I briefly reviewed your pull request. However, there are some bugs you must correct for this repository. Please make sure your code works on this repository when you request it from somebody. Thank you for your understanding.

------------------------------------------------------------------------------
D_Hydro [1747154235.132912] <anonymous> >> d_hydro shutting down normally
Running MATLAB script...
Opening Delft3D trim file: ../trim-45x45.dat
Unrecognized function or variable '**vs_use**'.

Error in waterlevel_delft3d4 (line 18)
trim = vs_use(['../trim-', name_output, '.dat'], 'quiet');
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error in run (line 112)
evalin('caller', strcat(scriptStem, ';'));
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Running WEADS_delft3d4.py...

CondaError: Run 'conda init' before 'conda activate'



LAUNCH: Launching script!

Input parameters for MEM

jinikeda commented on WEADS_delft3d4.py:
```diff
@@ -0,0 +1,109 @@
+# Step 1: Extract tidal datums from CSV
+print("Calculating tidal datums from extracted CSV...")
+tidal_csv = "tidal_metrics.csv"
+calculate_tidal_metrics_from_csv('water_level_extracted.csv', output_csv=tidal_csv)
```
'water_level_extracted.csv' Think about user input style rather than the specified name here.

jinikeda commented on src_point/preprocessing_delft3d4.py:
```diff
@@ -0,0 +1,108 @@
+    hydroclass_index[z > mhw_array] = 0  # dry
+    hydroclass_index[z <= mlw_array] = 2  # submerged
+
+    hydroclass_label = np.array(['intertidal'] * len(z), dtype=object)
```
Keep a consistent name 0: land, 1: intertidal, 2: subtidal

jinikeda commented on src_point/mem_delft3d4.py:
Future recommentations. Merge into mem.py

jinikeda commented on env_delft3d4.yml:
Do we need to create two virtual env?
user/local/delft3d and wead_delft3d4_env

jinikeda replied:
I will merge them.

jinikeda reviewed (changes requested):
Please consider modifications for each comment.

jinikeda commented on src_point/postprocessing_delft3d4.py:
```diff
@@ -0,0 +1,136 @@
+        values = [float(val) for line in f if line.strip() for val in line.strip().split()]
+
+    total = len(values)
+    side = int(np.sqrt(total))
```
Modify infer_grid_shape_from_dep(dep_file) function. This function only works when NX & NY are the same. Most of the problem NX != NY

jinikeda replied:
Further request for modifying the function.  See the below change.

jinikeda commented on src_point/postprocessing_delft3d4.py:
```diff
@@ -0,0 +1,136 @@
+def update_delft3d_rgh(outputMEMFile, grid_shape, output_rgh_file):
+    df = pd.read_csv(outputMEMFile)
+
+    class_to_chezy = {
+        55: 65.0,   # land / upland
+        40: 20.0,   # subtidal or water
+        16: 45.0,   # low productivity
+        23: 35.0,   # medium productivity
+        32: 25.0,   # high productivity
+        128: 65.0   # nodata default
+    }
```
Delft3D can also use Manning input! 
see Eq(4.7)
![image](https://github.com/user-attachments/assets/8f2247a5-f4ab-4ef2-bbe5-d78b91900e26)

jinikeda reviewed (changes requested):
Request to modify the codes.

jinikeda commented on WEADS_delft3d4.py:
Try to keep the previous history!

jinikeda commented on src_delft3d4/delft3d4_rgh.py:
```diff
@@ -0,0 +1,44 @@
+
+import numpy as np
+
+def create_rgh_from_dep(dep_filename, rgh_filename='output.rgh', roughness_value=22.0):
+    with open(dep_filename, 'r') as f:
+        lines = f.readlines()
```
Change the value of roughness_value to Manning's n
Also, add what is roughness_value.

jinikeda commented on src_delft3d4/delft3d4_rgh.py:
```diff
@@ -0,0 +1,44 @@
+
+import numpy as np
+
+def create_rgh_from_dep(dep_filename, rgh_filename='output.rgh', roughness_value=22.0):
```
I disagree with reading M and N sizes from the depth file. Use line 7 in a grid file!

jinikeda commented on src_point/general_functions_delft3d4.py:
```diff
@@ -0,0 +1,41 @@
+    return np.array(values)
+
+
+def infer_grid_shape_from_dep(dep_file):
+    """
+    Infers the shape (nx, ny) of the grid from the .dep file, assuming either a square
+    or rectangular layout. Tries to guess the shape based on total number of values.
+    """
+    values = read_dep_file(dep_file)
+    total = len(values)
+
+    # First try perfect square
+    side = math.isqrt(total)
+    if side * side == total:
+        return side, side
+
+    # Then try rectangular factors
+    for ny in range(1, total + 1):
+        if total % ny == 0:
+            nx = total // ny
+            if nx * ny == total:
+                return nx, ny
+
+    raise ValueError(f"Cannot infer grid shape from {dep_file} with {total} values.")
```
As I mentioned in the delft3d4_rgh.py
I disagree with reading M and N sizes from the depth file. Use line 7 in a grid file!

So, modify_dep_file function like 
values, rows, columns = read_dep_file(dep_file)
here 
values, _, _, = read_dep_file(dep_file) are also available. Here_ is a common convention to indicate that a return value is intentionally ignored.

jinikeda commented on src_point/preprocessing_delft3d4.py:
```diff
@@ -0,0 +1,106 @@
+    hydroclass_label[hydroclass_index == 0] = 'land'
+    hydroclass_label[hydroclass_index == 2] = 'subtidal'
+
+    # --- Fill manning and other values ---
+    mannings_n = np.full_like(z, 0.035, dtype=float)
+
+    df_out = pd.DataFrame({
```
Modify line 90. This value should not be explicitly provided. Read rgh file and then set the mannings_n.

jinikeda reviewed (changes requested):
1.  Modify the function of **infer_grid_shape_from_dep(dep_file)** in the **src_point/general_functions_delft3d4.py** by reading actual NX and NY numbers
2. Modify line 90 in the **src_point/preprocessing.py**
Reading manning n values from rgh file.

jinikeda commented on src_point/tidaldatums_delft3d4.py:
The code has bugs. Think about why x and y of tidal_metrics.csv and ecology.csv are empty? 
grid_coords_extracted.csv is acceptable, butI highly recommend making a function to read the grd file using Python.

shabnamm7 commented:
Thanks for the detailed review! I've addressed all requested changes in today’s updates:

1. Replaced the hardcoded `roughness_value=22.0` with a proper Manning’s n value (`0.02`) and clarified the docstring in `delft3d4_rgh.py`.
2. Implemented a function to read `NX` and `NY` from line 7 of the `.grd` file, as recommended. This removes reliance on `.dep` for shape inference.
3. Refactored the preprocessing script to **read Manning’s n values from the `.rgh` file**, instead of assigning a constant value.
4. Fixed the bug in `tidaldatums_delft3d4.py` where `x` and `y` were missing by properly aligning `pt` labels from the CSV.
5. Created and used a clean `read_grid_shape_from_grd()` function to improve consistency and robustness in shape handling.
6.  Also updated the `.dep` handling logic in `delft3d4_dep.py` to use the grid shape from the `.grd` file rather than guessing shape from values. This improves reliability across non-square domains.

Let me know if any other refinements are needed — I’m happy to adjust!

jinikeda reviewed (approved):
Ready to merge processes
